packages/web/lib/browser.js

```diff
@@ -119,30 +137,69 @@ const getBrowserScope = (parent) => ({
-          query: match.query,
-        });
-
-        // Navigate to the page
-        pageScope.internal.page = await browser.newPage();
-        await pageScope.internal.page.goto(href);
-
-        // TODO: support multple matching definitions
-        await evalRuleSet(match.ruleSet, pageScope);
+      //Get the nearest browser scope
+      const nearestBrowserScope = resolveInternalScope("browser", scope);
 
-        // TODO: check if the url has changed? If so, recurse and execute the necessary `page` rule
-
-        // Finally, close the page
-        pageScope.internal.page.close();
+      //If the nearest browser scope doesn't have a browser, create one
+      let browser = nearestBrowserScope.internal.browser;
+      if (!browser) {
+        browser = nearestBrowserScope.internal.browser = await puppeteer.launch(
+          {
+            headless: false,
+          }
+        );
       }
 
-      return href;
+      if (match) {
+        await Promise.all(
+          match.ruleSets.map(async (ruleSet) => {
+            //Create a new page scope in the same scope that the page rule was created in and navigate to href
+            const newPageScope = await preparePage(match.parent)(browser, href);
+
+            // inject args and "url" as variables into the new page scope
+            Object.assign(newPageScope.vars, match.path, {
+              url: href,
+              hash: match.hash,
+              query: match.query,
+            });
+
+            await evalRuleSet(ruleSet, newPageScope);
+
+            // TODO: check if the url has changed? If so, recurse and execute the necessary `page` rule
+            const data = newPageScope.internal.data;
+            if (Object.keys(data).length) {
+              if (data.url) {
+                console.warn(
+                  "WARNING:: Setting the url key using a data function will override the default url key (See docs <https://....>)"
+                );
+              }
+              data.url = href;
+              //Grabs the nearest config where output is defined, else return false
+              const config = resolveInternal(
+                "config",
+                newPageScope,
+                (config) => !!config.output,
+                false
+              );
+              if (config) {
+                config.writeStream.write(JSON.stringify(data) + "\n", {
+                  flags: "a",
+                });
+                config.writeStream.end("");
+              } else {
+                console.log(JSON.stringify(data));
+              }
+            }
+            // TODO: check if the url has changed? If so, recurse and execute and necessary `pageDef` functions
+            // Finally, close the page
+            newPageScope.internal.page.close();
+          })
+        );
+      } else {
+        //Create a new page in the nearest browser scope and navigate to href
+        await preparePage(nearestBrowserScope)(browser, href);
+        return href;
```

Comment:
move this `return` after the `if ... else`
let's always return the `href` from a visit call

we don't ever want to return `undefined` which is what currently happens if a match was found

We can use typescript later to ensure this